perp/man/perpctl.8
.\" perpctl.8
.\" wcm, 2009.12.01 - 2011.02.01
.\" ===
.TH perpctl 8 "March 2011" "perp-2.04" "persistent process supervision"
.de ZZ
.  br
.  ns
..
.SH NAME
perpctl \- runtime control utility for
.BR perpd (8)
services
.SH SYNOPSIS
.B perpctl [-hV] [-b
.I basedir
.B ] [-g] [-L] [-q]
.I cmd sv
.B [
.I sv ...
.B ]
.SH DESCRIPTION
.B perpctl
sends the command specified in
.I cmd
to the
.BR perpd (8)
control interface for each service argument
.IR sv .
.PP
The argument
.I cmd
may be given as a single word of any length,
but only the first letter is considered.
The available commands include
(with mnemonic in parentheses):
.PP
.B A
(Activate)
.RS
Sets the sticky bit on each
.I sv
service directory argument and
sends SIGHUP to
.BR perpd (8).
The effect is to activate each
.I sv
service.
.RE
.PP
.B X
([e]Xit)
.RS
Unsets the sticky bit on each
.I sv
service directory argument and
sends SIGHUP to
.BR perpd (8).
The effect is to bring down each
.I sv
service and remove it
from the set of services
.BR perpd (8)
is monitoring.
.RE
.PP
.B d
(down)
.RS
If the service is running,
send it a sequence of SIGTERM and SIGCONT signals to bring it down.
.BR perpd (8)
will flag the service as wanting down:
if the service stops it will not be restarted. 
.RE
.PP
.B u
(up)
.RS
If the service is not already running, start it.
.BR perpd (8)
will flag the service as wanting up:
if the service stops it will be restarted.
.RE
.PP
.B o
(once)
.RS
If the service is not already running, start it.
.BR perpd (8)
will flag the service to run once:
if the service stops it will not be restarted.
.RE
.PP
.B p
(pause)
.RS
Send the service a SIGSTOP signal.
Normally this suspends execution of the service.
.BR perpd (8)
will flag the service as paused.
.RE
.PP
.B c
(continue)
.RS
Send the service a SIGCONT signal.
Normally a paused service will then resume execution.
.BR perpd (8)
will remove a pause flag on the service.
.RE
.PP
.B a
(alarm)
.ZZ
.B h
(hup)
.ZZ
.B i
(interrupt)
.ZZ
.B k
(kill)
.ZZ
.B t
(term)
.ZZ
.B w
(winch)
.ZZ
.B 1
(usr1)
.ZZ
.B 2
(usr2)
.RS
Send the service a corresponding signal:
SIGALRM, SIGHUP, SIGINT, SIGKILL, SIGTERM, SIGWINCH, SIGUSR1 or SIGUSR2.
.RE
.PP
.B D
(meta-Down)
.ZZ
.B U
(meta-Up)
.RS
When given in upper-case, the
.B d
(down) and
.B u
(up) commands described above
are applied to both the main and log services.
.RE
.PP
The signal/control commands listed above
are applied to an active service process
running from the ``start'' target of its
.BR perpetrate (5)
runscript.
.BR perpd (8)
will otherwise ignore any of the commands described above
if received while a service is resetting,
except for the commands
.B c
(continue/SIGCONT)
or
.B k
(kill/SIGKILL).
.SH OPTIONS
.TP
.B \-b basedir
Base directory.
Sets the base directory containing the
.I sv
arguments.
If not set,
.B perpctl
will look for a value set in the environmental variable PERP_BASE.
If neither of these is set,
.B perpctl
will operate on the current working directory.
.TP
.B \-g
Group.
Apply the requested command to the process group id (pgid) of each
.I sv
service.
Normally the signal is applied only to the single process id of the service.
.BR perpd (8)
runs each main and log process in its own separate process group;
the
.B \-g
option directs
.BR perpd (8)
to signal all process running with the pgid of the service.
May be combined with the
.B \-L
option to signal all processes in the process group of the logging service.
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-L
Logger.
Apply the
.I cmd
argument to the active logging service found in
.I sv
services.
By default, the
.I cmd
argument is applied to the main service.
The
.B \-L
option is not allowed when using any of the meta-commands
.BR D ", " U ", or " X .
.TP
.B \-q
Quiet.
Normally after successfully applying the
.I cmd
to each
.IR sv ,
.B perpctl
reports a brief message to stderr.
The
.B \-q
option may be used to suppress these messages.
.TP
.B \-V
Version.
Print the version number of the program to stderr and exit.
.SH DIAGNOSTICS
For each
.I sv
successfully processed,
.B perpctl
prints a line to stderr in the form:
.IP
.IR sv :
ok
.PP
For each
.I sv
not successfully processed,
.B perpctl
prints a brief diagnostic to stderr and continues processing any remaining
.IR sv .
.SH EXIT STATUS
.PP
.B perpctl
exits with one of the following values:
.TP
0
Success.
The
.I cmd
was successfully delivered to all
.I sv
service arguments.
.TP
100
Usage error.
For unknown options, missing arguments, or other command-line errors.
Prints a brief diagnostic message to stderr on exit.
.TP
111
System error.
One or more errors were encountered while processing.
These may include unexpected failures of system calls,
privilege and/or resource problems,
or configuration errors in the base directory.
A brief diagnostic message is printed to stderr for each error encountered.
.SH AUTHOR
Wayne Marshall, http://b0llix.net/perp/
.SH SEE ALSO
.nh
.BR perp_intro (8),
.BR perpboot (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perphup (8),
.BR perpls (8),
.BR perpok (8),
.BR perpstat (8),
.BR sissylog (8),
.BR tinylog (8)
.\" eof


CHANGES
# CHANGES
# perp release history and changelog
# wcm, 2010.02.17 - 2011.03.22
# ===

perp-2.04 (2011.03.22):
 * perphup:
   - added option -t, trigger perpd shutdown
 * perpboot:
   - waitpid() for deux[1] on SIGTERM
 * perp-setup:
   - added arg for ctldir
   - configure symlink /etc/perp/.control --> ctldir
   - configure rc.perp bootscript for runchoom(8)
 * runchoom:
   - new, utility for linux "oom killer" abatement

perp-2.03 (2011.03.11):
 * perp-setup:
   - in rc.perp boot script for perpd(8), setup default runlimit(8)
     configuration to hard limits for RLIMIT_NPROC, RLIMIT_NOFILE

perp-2.02 (2011.02.22):
 * perpd:
   - eliminated some spurious/redundant client connection logging

perp-2.01 (2011.02.11):
 * perp-setup:
   - runlimit added to .boot/rc.perp script
 * perpd:
   - added option -g, enable group permission on control socket
   - additional comments in LIMITS section of perpd(8) manual
 * runlimit:
   - added option -E, take resource limits from environment
   - added option -F, take resource limits from file
 * lasagna:
   - outvec/mk_outvec.c: portability fix for dietlibc

perp-2.00 (2011.02.02):
 * perpd:
   - coalesce supervisor (formerly perpetrate(8)) into scanner/controller
   - runs as a single process
   - all ipc with client applications via single domain socket 
   - ipc protocol version 2 (see PROTO_V2.txt)
 * perpetrate:
   - eliminate perpetrate(8) executable (subsumed by perpd(8))
   - perpetrate(5) manual page retained for runscript conventions
 * perpctl:
   - add commands [A | X], activate/deactivate service
 * libasagna:
   - nuscan.h, improve numeric scanning interface (supercedes nscan.h)
   - other iterative improvements incorporated from projects hdb and ezcdb
 
perp-0.00 (2010.01.12):
 - maiden public release

### EOF


perp/perpctl.c
/* perpctl.c
** perpctl: perpetrate supervisor control interface
** wcm, 2008.01.23 - 2010.01.05
** ===
*/

/* standard library headers: */
#include <stdlib.h>

/* unix headers: */
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>

/* lasagna headers: */
#include "buf.h"
#include "fd.h"
#include "nextopt.h"
#include "padlock.h"
#include "pidlock.h"
#include "pkt.h"
#include "sig.h"
#include "sysstr.h"
#include "uchar.h"
#include "upak.h"

/* local headers: */
#define  EPUTS_OUTVEC  1
#include "eputs.h"
#include "perp_common.h"
#include "perplib.h"

static const char *progname = NULL;
static const char prog_usage[] =
    " [-hV] [-b basedir] [-L] [DUXduopcahikqtw12] sv [sv ...]";


/* prototypes in scope: */
static int     do_control(const char *svdir, const pkt_t pkt);


/* do_control():
** for service named in svdir
**   send command packet to input fifo
**   read response packet from output fifo
**
** notes:
**   cwd is preset to service control directory
**
** return:
**   0 : ok
**  -1 : error, diagnostic on stderr
*/
static
int do_control(const char *svdir, const pkt_t pkt)
{
   int            ctl_lock;
   int            fifo_0, fifo_1;
   pkt_t          reply;
   int            n;
   int            terrno;

   /* open/acquire client access lock: */
   ctl_lock = open(CTL_LOCK, O_WRONLY);
   if (ctl_lock == -1) {
      eputs(svdir, ": failure on open() for client lock ", CTL_LOCK, ": ",
            sysstr_errno(errno));
      return -1;
   }
   if (padlock_exlock(ctl_lock, PADLOCK_WAIT) == -1) {
      eputs(svdir, ": failure acquiring lock on client lock ", CTL_LOCK, ": ",
            sysstr_errno(errno));
      close(ctl_lock);
      return -1;
   }

   /* open control fifos: */
   fifo_0 = open(CTL_IN, O_WRONLY);
   if (fifo_0 == -1) {
      eputs(svdir, ": failure on open() for input fifo ", CTL_IN, ": ",
            sysstr_errno(errno));
      close(ctl_lock);
      return -1;
   }
   fifo_1 = open(CTL_OUT, O_RDONLY);
   if (fifo_1 == -1) {
      eputs(svdir, ": failure on open for output fifo ", CTL_OUT, ": ",
            sysstr_errno(errno));
      close(fifo_0);
      close(ctl_lock);
      return -1;
   }

   /* send command packet: */
   n = pkt_write(fifo_0, pkt);
   if (n == -1) {
      int            terrno = errno;
      eputs(svdir, ": failure packet_write() for command packet: ",
            sysstr_errno(terrno));
      close(fifo_1);
      close(fifo_0);
      close(ctl_lock);
      return -1;
   }

   /* get reply packet: */
   n = pkt_read(fifo_1, reply);

   /* done with connection: */
   terrno = errno;
   close(fifo_1);
   close(fifo_0);
   close(ctl_lock);
   errno = terrno;

   if (n == -1) {
      eputs(svdir, ": failure on pkt_read() for command reply: ",
            sysstr_errno(errno));
      return -1;
   }

   if (pkt_proto(reply) != 1) {
      eputs(svdir, ": protocol error found in server response");
      return -1;
   }

   if ((pkt_type(reply) == 'E') && (pkt_size(reply) == 4)) {
      uchar_t       *payload = pkt_data(reply);
      int            err = (int) upak32_unpack(payload);

      /* error reported in reply packet? */
      if (err != 0) {
         eputs(svdir, ": server reply to command packet with error: ",
               sysstr_errno(err));
         return -1;
      }
   } else {
      eputs(svdir, ": protocol error found in server repsonse");
      return -1;
   }

   /* success: */
   eputs(svdir, ": ok");
   return 0;
}


int main(int argc, char *argv[])
{
   char           opt;
   nextopt_t      nopt = nextopt_INIT(argc, argv, ":hVb:L");
   int            logflag = 0;
   pkt_t          pkt = pkt_INIT(1, 'C', 1);    /* a "command" packet */
   uchar_t       *cmd = pkt_data(pkt);
   const char    *basedir = NULL;
   int            fd_base;
   int            errs = 0;

   progname = nextopt_progname(&nopt);
   while ((opt = nextopt(&nopt))) {
      char           optc[2] = { nopt.opt_got, '\0' };
      switch (opt) {
      case 'h':
         usage();
         die(0);
         break;
      case 'V':
         version();
         die(0);
         break;
      case 'b':
         basedir = nopt.opt_arg;
         break;
      case 'L':
         logflag = 1;
         break;
      case ':':
         fatal_usage("missing argument for option -", optc);
         break;
      case '?':
         if (nopt.opt_got != '?') {
            fatal_usage("invalid option: -", optc);
         }
         /* else fallthrough: */
      default:
         die_usage();
         break;
      }
   }

   argc -= nopt.arg_ndx;
   argv += nopt.arg_ndx;

   if (!*argv) {
      eputs(progname, ": usage error: missing arguments");
      die_usage();
   }

   /* next argv is the control command (can be taken from first letter): */
   switch (cmd[0] = *argv[0]) {
   case 'd':
   case 'u':
   case 'o':
   case 'p':
   case 'c':
   case 'a':
   case 'h':
   case 'i':
   case 'k':
   case 'q':
   case 't':
   case 'w':
   case '1':
   case '2':
      if (logflag) {
         /* control command for log service: */
         eputs("shifting command for log service");
         cmd[0] += 0x7f;
      }
      break;
   case 'D':
   case 'U':
   case 'X':
      if (logflag) {
         fatal_usage("meta-command '", *argv,
                     "' may not be used with option -L");
      }
      break;
   default:
      fatal_usage("unknown control command '", *argv, "'");
      break;
   }

   --argc;
   ++argv;
   if (!*argv) {
      fatal_usage("missing argument: no service(s) specified");
   }
#if 0
   {
      cmd[1] = '\0';
      eputs(progname, ": sending command ", cmd);
   }
#endif

   if (basedir == NULL)
      basedir = getenv("PERP_BASE");
   if ((basedir == NULL) || (basedir[0] == '\0'))
      basedir = ".";

   if (chdir(basedir) != 0) {
      fatal_syserr("unable to chdir() to ", basedir);
   }

   if ((fd_base = open(".", O_RDONLY | O_NONBLOCK)) == -1) {
      fatal_syserr("unable to open() on ", basedir);
   }

   /* ignore SIGPIPE: */
   sig_ignore(SIGPIPE);

   /* loop through service directory argument(s) and send control packet: */
   for ( /* */ ; *argv != NULL; ++argv) {
      struct stat    st;
      const char    *ctlpath;
      int            e;

      if (stat(*argv, &st) != 0) {
         eputs(*argv, ": failure stat() on service directory: ",
               sysstr_errno(errno));
         ++errs;
         continue;
      }
      if (!S_ISDIR(st.st_mode)) {
         eputs(*argv, ": not a directory");
         ++errs;
         continue;
      }

      ctlpath = perp_ctlpath(&st);
      if (perp_ready(ctlpath) == -1) {
         if ((errno == ENXIO) || (errno == ENOENT)) {
            eputs(*argv, ": supervisor not running");
         } else {
            eputs(*argv, ": failure checking supervisor: ",
                  sysstr_errno(errno));
         }
         ++errs;
         continue;
      }

      if (chdir(ctlpath) == -1) {
         eputs(*argv, ": failure chdir() to service control directory: ",
               sysstr_errno(errno));
         ++errs;
         continue;
      }

      /* the business: */
      e = do_control(*argv, pkt);
      if (e != 0)
         ++errs;
      if (fchdir(fd_base) == -1) {
         fatal_syserr("failure fchdir() to base directory");
      }
   }

   return (errs ? 111 : 0);
}


/* eof (perpctl.c) */
